frame/contracts/src/wasm/runtime.rs

```diff
@@ -705,26 +729,127 @@ where
+	fn clear_storage(&mut self, key_type: KeyType, key_ptr: u32) -> Result<u32, TrapReason> {
 		let charged = self.charge_gas(RuntimeCosts::ClearStorage(self.ext.max_value_size()))?;
-		let mut key: StorageKey = [0; 32];
-		self.read_sandbox_memory_into_buf(key_ptr, &mut key)?;
-		let outcome = self.ext.set_storage(key, None, false)?;
+		let key = self.read_sandbox_memory(key_ptr, key_type.len::<E::T>()?)?;
+		let outcome = match key_type {
+			KeyType::Fix => self.ext.set_storage(
+				&FixSizedKey::try_from(key).map_err(|_| Error::<E::T>::DecodingFailed)?,
+				None,
+				false,
+			)?,
+			KeyType::Variable(_) => self.ext.set_storage_transparent(
+				&VarSizedKey::<E::T>::try_from(key).map_err(|_| Error::<E::T>::DecodingFailed)?,
+				None,
+				false,
+			)?,
+		};
+
 		self.adjust_gas(charged, RuntimeCosts::ClearStorage(outcome.old_len()));
 		Ok(outcome.old_len_with_sentinel())
 	}
 
+	fn get_storage(
+		&mut self,
+		key_type: KeyType,
+		key_ptr: u32,
+		out_ptr: u32,
+		out_len_ptr: u32,
+	) -> Result<ReturnCode, TrapReason> {
+		let charged = self.charge_gas(RuntimeCosts::GetStorage(self.ext.max_value_size()))?;
+		let key = self.read_sandbox_memory(key_ptr, key_type.len::<E::T>()?)?;
+		match key_type {
+			KeyType::Fix =>
+				if let Some(value) = self.ext.get_storage(
+					&FixSizedKey::try_from(key).map_err(|_| Error::<E::T>::DecodingFailed)?,
+				) {
+					self.adjust_gas(charged, RuntimeCosts::GetStorage(value.len() as u32));
+					self.write_sandbox_output(
+						out_ptr,
+						out_len_ptr,
+						&value,
+						false,
+						already_charged,
+					)?;
+					Ok(ReturnCode::Success)
+				} else {
+					self.adjust_gas(charged, RuntimeCosts::GetStorage(0));
+					Ok(ReturnCode::KeyNotFound)
+				},
+			KeyType::Variable(_) => {
+				if let Some(value) = self.ext.get_storage_transparent(
+					&VarSizedKey::<E::T>::try_from(key)
+						.map_err(|_| Error::<E::T>::DecodingFailed)?,
+				) {
+					self.adjust_gas(charged, RuntimeCosts::GetStorage(value.len() as u32));
+					self.write_sandbox_output(
+						out_ptr,
+						out_len_ptr,
+						&value,
+						false,
+						already_charged,
+					)?;
+					Ok(ReturnCode::Success)
+				} else {
+					self.adjust_gas(charged, RuntimeCosts::GetStorage(0));
+					Ok(ReturnCode::KeyNotFound)
+				}
+			},
+		}
+	}
+
+	fn contains_storage(&mut self, key_type: KeyType, key_ptr: u32) -> Result<u32, TrapReason> {
+		let charged = self.charge_gas(RuntimeCosts::ContainsStorage(self.ext.max_value_size()))?;
+		let key = self.read_sandbox_memory(key_ptr, key_type.len::<E::T>()?)?;
+		match key_type {
+			KeyType::Fix => {
+				if let Some(len) = self.ext.get_storage_size(
+					&FixSizedKey::try_from(key).map_err(|_| Error::<E::T>::DecodingFailed)?,
+				) {
+					self.adjust_gas(charged, RuntimeCosts::ContainsStorage(len));
+					Ok(len)
+				} else {
+					self.adjust_gas(charged, RuntimeCosts::ContainsStorage(0));
+					Ok(SENTINEL)
+				}
```

Comment:
This part seems to be the same for both branches. Can't it be factored out as in `clear_storage`?